### DISCUSSION TITLE
How to automatically exclude certain facet values in search?

### DISCUSSION THREAD

**Discussion by akaustav** [coveo/cli#480] [Q&A] [answered]
Hi,

We are converting a classic `coveo-search-ui` application to use `@coveo/headless` and `@coveo/angular`. In the original application (which used `coveo-search-ui`) there was requirements to:

1. not allow any manual filtering based on certain criteria:
    1. no UI checkboxes to select/deselect, and
    2. advanced keyword searches using [special field operators](https://docs.coveo.com/en/1897/searching-with-coveo/advanced-field-queries) was disabled.
1. filter out / exclude all PDF files from the result list every time the user searched for a specific word / phrase.

This was achieved by adding an `aq` field to the `application/x-www-form-urlencoded` request body on every search. For example, when searching for the term: `hello`, the request body would look like this (trimmed to show relevant fields and parsed into YAML format for better syntax highlighting):

```yaml
q: hello
aq: (@filetype<>pdf)
enableQuerySyntax: false
...
```

Is there a way to achieve this exact requirement using `@coveo/headless` and `@coveo/angular`?
The sample application generated by `@coveo/cli` demonstrates how to get facet values and select a specific value, but we are unable to find an appropriate method to achieve facet value exclusions.

Ameet

**Comment by akaustav** [comment]
From what we have gathered so far, there's two approaches / thought processes:
1. Exclude PDFs while crawling the website - so that they are not indexed to begin with
2. Exclude PDFs using facets - but we don't know see a straightforward way to achieve facet value exclusions

Which approach would the coveo team suggest and why? Any code examples / samples would be appreciated.

**Comment by y-lakhdar** [comment, marked as answer]
Hi Ameet,

I recommend moving the filtering logic to your pipeline instead of defining the `aq` and `enableQuerySyntax` in the UI.
![image](https://user-images.githubusercontent.com/12199712/134187293-8a6a2875-bdf5-454b-9b6f-bf9a6ce4982e.png)
![image](https://user-images.githubusercontent.com/12199712/134187324-ef27d139-90d7-4126-bc86-edfc9c8a39d7.png)

Moving logic to the pipeline will:
1. help you remove complexity from your front-end code
2. allow you to manipulate your search page results without having to re-deploy/edit your search page

As for the PDF exclusion, the suggestion depends on the intention.
If you don’t want to expose PDFs in your search page, I recommend excluding them during the indexation.
However, if you only want to filter them out from the result list when the user has entered a query, then I recommend using a pipeline condition as shown in the second screenshot.

Does that answer your question?

Useful links:
* [Manage Filter Rules](https://docs.coveo.com/en/3410/tune-relevance/manage-filter-rules)
* [Manage Query Parameter Rules](https://docs.coveo.com/en/3411/tune-relevance/manage-query-parameter-rules)

**Reply by akaustav** [reply]
Thanks, @y-lakhdar. We'll try moving the filtering logic to our pipeline.

**Reply by akaustav** [reply]
@y-lakhdar - When do these filter and query parameter rules take effect? Do they take effect immediately or maybe after the next scheduled crawl?

**Reply by y-lakhdar** [reply]
They will take effect immediately.

**Reply by akaustav** [reply]
Thanks, @y-lakhdar. I was actually modifying the settings in a different organization and not seeing any differences in results. Hence, my earlier question. The pipeline filters worked perfectly for us - once I put those into the correct Coveo organization.
